spinalcordtoolbox/moco.py

```diff
@@ -0,0 +1,855 @@
+
+    # Delete temporary files
+    if param.remove_temp_files == 1:
+        sct.printv('\nDelete temporary files...', param.verbose)
+        sct.rmtree(path_tmp, verbose=param.verbose)
+
+    # come back to working directory
+    os.chdir(curdir)
+
+    # display elapsed time
+    elapsed_time = time.time() - start_time
+    sct.printv('\nFinished! Elapsed time: ' + str(int(np.round(elapsed_time))) + 's', param.verbose)
+
+    sct.display_viewer_syntax(
+        [os.path.join(param.path_out, sct.add_suffix(os.path.basename(param.fname_data), param.suffix)),
+         param.fname_data], mode='ortho,ortho')
+
+
+def moco(param):
+    """
+    Main function that performs motion correction.
+    :param param:
+    :return:
+    """
+    # retrieve parameters
+    file_data = param.file_data
+    file_target = param.file_target
+    folder_mat = param.mat_moco  # output folder of mat file
+    todo = param.todo
+    suffix = param.suffix
+    verbose = param.verbose
+
+    # other parameters
+    file_mask = 'mask.nii'
+
+    sct.printv('\nInput parameters:', param.verbose)
+    sct.printv('  Input file ............ ' + file_data, param.verbose)
+    sct.printv('  Reference file ........ ' + file_target, param.verbose)
+    sct.printv('  Polynomial degree ..... ' + param.poly, param.verbose)
+    sct.printv('  Smoothing kernel ...... ' + param.smooth, param.verbose)
+    sct.printv('  Gradient step ......... ' + param.gradStep, param.verbose)
+    sct.printv('  Metric ................ ' + param.metric, param.verbose)
+    sct.printv('  Sampling .............. ' + param.sampling, param.verbose)
+    sct.printv('  Todo .................. ' + todo, param.verbose)
+    sct.printv('  Mask  ................. ' + param.fname_mask, param.verbose)
+    sct.printv('  Output mat folder ..... ' + folder_mat, param.verbose)
+
+    # create folder for mat files
+    sct.create_folder(folder_mat)
+
+    # Get size of data
+    sct.printv('\nData dimensions:', verbose)
+    im_data = Image(param.file_data)
+    nx, ny, nz, nt, px, py, pz, pt = im_data.dim
+    sct.printv(('  ' + str(nx) + ' x ' + str(ny) + ' x ' + str(nz) + ' x ' + str(nt)), verbose)
+
+    # copy file_target to a temporary file
+    sct.printv('\nCopy file_target to a temporary file...', verbose)
+    file_target = "target.nii.gz"
+    convert(param.file_target, file_target, verbose=0)
+
+    # Check if user specified a mask
+    if not param.fname_mask == '':
+        # Check if this mask is soft (i.e., non-binary, such as a Gaussian mask)
+        im_mask = Image(param.fname_mask)
+        if not np.array_equal(im_mask.data, im_mask.data.astype(bool)):
+            # If it is a soft mask, multiply the target by the soft mask.
+            im = Image(file_target)
+            im_masked = im.copy()
+            im_masked.data = im.data * im_mask.data
+            im_masked.save(verbose=0)  # silence warning about file overwritting
```

Comment:
Original application of this logic in moco

Reply:
thank you for finding the origin of this code